### PR TITLE
Integrate `scanResultBlob` pattern on `robonode-client` and `bioauth-flow-rpc` 

### PR DESCRIPTION
**Pull request by dmitrylavrenov**
Should be merged after #1065 

As agreed integration of `scanResultBlob` pattern on `bioauth-flow-rpc` and `robonode-client` sides should be backward compatible.

### CODE REVIEW

File: crates/bioauth-flow-rpc/src/lib.rs [hunk 1]
```diff
@@ -325,12 +348,17 @@ where
             .await
             .map_err(EnrollError::Robonode)?;
 
+        let scan_result_blob = response.scan_result_blob.clone();
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/lib.rs:351]: Why clone?

Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/lib.rs]: https://github.com/humanode-network/humanode/pull/1026/commits/68a682de76ccf28f18c3e478c348a52c8c384bfc

File: crates/robonode-server/src/logic/op_authenticate.rs [hunk 2]
```diff
@@ -109,6 +112,8 @@ where
             .await
             .map_err(Error::InternalErrorEnrollment)?;
 
+        let scan_result_blob = enroll_res.scan_result_blob.clone();
```
Comment (MOZGIII) [inline crates/robonode-server/src/logic/op_authenticate.rs:115]: Why clone?

File: crates/bioauth-flow-rpc/src/lib.rs [hunk 3]
```diff
@@ -315,7 +338,7 @@ where
         let (opaque_liveness_data, signature) = self.sign(public_key.clone(), &liveness_data).await
             .map_err(EnrollError::Sign)?;
 
-        self.robonode_client
+        let response = self.robonode_client
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/lib.rs:341]: We might want to unpack the result here (and in similar places) right away

File: crates/robonode-client/src/enroll.rs [hunk 4]
```diff
@@ -118,11 +130,17 @@ mod tests {
             public_key: b"123",
             liveness_data_signature: b"signature",
         };
+        let sample_response = serde_json::json!({
+            "scanResultBlob": "scanResultBlob"
+        });
+
+        let expected_response: EnrollResponse =
+            serde_json::from_value(sample_response.clone()).unwrap();
```
[the remark this answers is not in the document]
Comment (dmitrylavrenov) [reply crates/robonode-client/src/enroll.rs:138]: Tests, used later to be passed to `enroll` request.

File: crates/robonode-server/src/logic/op_authenticate.rs (hunk 2 above)
Comment (dmitrylavrenov) [reply crates/robonode-server/src/logic/op_authenticate.rs:115]: `enroll_res` has dropped before returning the result.

File: crates/bioauth-flow-rpc/src/lib.rs (hunk 3 above)
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/lib.rs]: https://github.com/humanode-network/humanode/pull/1026/commits/6ba5df193d4b034dcd274331f2b718bb85014438

File: crates/robonode-server/src/logic/op_authenticate.rs (hunk 2 above)
Comment (MOZGIII) [reply crates/robonode-server/src/logic/op_authenticate.rs:115]: Well, exactly - so why not deconstruct it into pieces? Take the `scan_result_blob` right from there.

Comment (MOZGIII) [reply crates/robonode-server/src/logic/op_authenticate.rs:115]: Note that even without manual destructuring, if you were to just from the clone here rust is smart enough to figure out what to do here and take the `scan_result_blob` out without cloning.

File: crates/bioauth-flow-rpc/src/lib.rs [hunk 5]
```diff
@@ -353,7 +383,7 @@ where
 
         info!("Bioauth flow - authentication complete");
 
-        info!(message = "We've obtained an auth ticket", auth_ticket = ?response.auth_ticket);
+        info!(message = "We've obtained an auth ticket", auth_ticket = ?auth_ticket);
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/lib.rs:386]:
```suggestion
        info!(message = "We've obtained an auth ticket", ?auth_ticket);
```

**Comment by MOZGIII** [comment]
This needs to be tested against the corresponding web app implementation before it can be merged. The reason is we don't know if this would work at all.

File: crates/robonode-server/src/logic/op_authenticate.rs (hunk 2 above)
Comment (dmitrylavrenov) [reply crates/robonode-server/src/logic/op_authenticate.rs]: Got it, improved - https://github.com/humanode-network/humanode/pull/1026/commits/78dd8e438226aa986178e235f8af4e83894b023d.

**Comment by dmitrylavrenov** [comment]
> This needs to be tested against the corresponding web app implementation before it can be merged. The reason is we don't know if this would work at all.

Sure, will check.

**Comment by dmitrylavrenov** [comment]
> This needs to be tested against the corresponding web app implementation before it can be merged. The reason is we don't know if this would work at all.

Just tested, works with it6.
image

**Comment by MOZGIII** [comment]
This PR seems to include both changes for robonode and for the chain, please make it draft and update the base

**Comment by dmitrylavrenov** [comment]
> This PR seems to include both changes for robonode and for the chain, please make it draft and update the base

Sure, will make it draft. Just robonode side is here - https://github.com/humanode-network/humanode/pull/1037.

File: crates/bioauth-flow-rpc/src/errors/robonode_request.rs
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/errors/robonode_request.rs:1]: Not a big fan the name for how this is structured in terms of naming

File: crates/robonode-client/src/authenticate.rs [hunk 6]
```diff
@@ -54,16 +57,29 @@ pub enum AuthenticateError {
     /// The person was not found, it is likely because they haven't enrolled first.
     #[error("person not found")]
     PersonNotFound,
+    /// The person was not found, it is likely because they haven't enrolled first, returned blob.
+    #[error("person not found, returned blob")]
+    PersonNotFoundReturnedBlob(ScanResultBlob),
```
Comment (MOZGIII) [inline crates/robonode-client/src/authenticate.rs:62]: I suggest we call this one `PersonNotFound(ScanResultBlob),`, and rename the old one to `PersonNotFoundNoBlob`

Comment (MOZGIII) [reply crates/robonode-client/src/authenticate.rs:62]: Do we still plan to support the old robonodes?

Comment (dmitrylavrenov) [reply crates/robonode-client/src/authenticate.rs:62]: Right, forward compatibility on `robonode-server` side, but back-forward compatibility on `robonode-client` side.

File: crates/bioauth-flow-rpc/src/errors/robonode_request.rs
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/errors/robonode_request.rs:1]: Got it, could you please suggest a better one ?

File: crates/robonode-client/src/authenticate.rs (hunk 6 above)
Comment (dmitrylavrenov) [reply crates/robonode-client/src/authenticate.rs]: https://github.com/humanode-network/humanode/pull/1026/commits/d9e6b0f831ac9ca3752f4e8716364847fba33cad

File: crates/bioauth-flow-rpc/src/errors/authenticate.rs [hunk 7]
```diff
@@ -4,18 +4,14 @@ use rpc_validator_key_logic::Error as ValidatorKeyError;
 use sp_api::ApiError;
 use sp_runtime::transaction_validity::InvalidTransaction;
 
-use super::{api_error_code, sign::Error as SignError};
+use super::{api_error_code, robonode_request::Error as RobonodeRequestError};
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/errors/authenticate.rs:7]: We have a lot of items to keep track of as is, let's minimize the aliases like `RobonodeRequestError` and use `robonode_request::Error` directly here and everywhere else

File: crates/bioauth-flow-rpc/src/errors/mod.rs [hunk 8]
```diff
@@ -1,9 +1,12 @@
 //! All bioauth flow error kinds that we expose in the RPC.
 
 pub mod authenticate;
+pub mod authenticate_v2;
 pub mod enroll;
+pub mod enroll_v2;
 pub mod get_facetec_device_sdk_params;
 pub mod get_facetec_session_token;
+pub mod robonode_request;
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/errors/mod.rs:9]:
```suggestion
pub mod robonode;
```

File: crates/bioauth-flow-rpc/src/errors/robonode_request.rs [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+/// The robonode requests related error.
+#[derive(Debug)]
+pub enum Error<T: std::error::Error + 'static> {
+    /// An error that can occur during validator key extraction.
+    KeyExtraction(ValidatorKeyError),
+    /// An error that can occur during signing process.
+    Sign(SignError),
+    /// An error that can occur during doing a call into robonode.
+    Robonode(robonode_client::Error<T>),
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/errors/robonode_request.rs:15]:
```suggestion
    Client(robonode_client::Error<T>),
```

File: crates/bioauth-flow-rpc/src/errors/robonode_request.rs [hunk 10]
```diff
@@ -0,0 +1,16 @@
+//! The robonode requests related error.
+
+use rpc_validator_key_logic::Error as ValidatorKeyError;
+
+use super::sign::Error as SignError;
+
+/// The robonode requests related error.
+#[derive(Debug)]
+pub enum Error<T: std::error::Error + 'static> {
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/errors/robonode_request.rs:9]:
```suggestion
pub enum FlowBaseError<T: std::error::Error + 'static> {
```
If we take https://github.com/humanode-network/humanode/pull/1026/files#r1719196297 into consideration

File: crates/bioauth-flow-rpc/src/errors/enroll_v2.rs [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+use super::{api_error_code, robonode_request::Error as RobonodeRequestError};
+use crate::error_data;
+
+/// The `enroll_v2` method error kinds.
+#[derive(Debug)]
+pub enum Error {
+    /// An error that can occur during doing a request to robonode.
+    RobonodeRequest(RobonodeRequestError<robonode_client::EnrollError>),
+}
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/errors/enroll_v2.rs:13]:
```suggestion
pub struct Error(pub shared::BaseFlowError<robonode_client::EnrollError>);
```

File: crates/robonode-client/src/authenticate.rs [hunk 12]
```diff
@@ -51,19 +54,32 @@ pub enum AuthenticateError {
     /// The provided liveness data was invalid.
     #[error("invalid liveness data")]
     InvalidLivenessData,
+    /// The person was not found, it is likely because they haven't enrolled first, no blob.
+    #[error("person not found, no blob")]
+    PersonNotFoundNoBlob,
```
Comment (MOZGIII) [inline crates/robonode-client/src/authenticate.rs:59]:
```suggestion
    /// The person was not found, it is likely because they haven't enrolled first, no blob.
    #[error("person not found")]
    PersonNotFoundNoBlob,
```

We can call those variants `NoBlob`, but we should avoid adding a different error message.

Also, to avoid confusion that causes this bikeshedding in the first place, I suggest that we add though comments on what is the reason for not having a blob here.

We *must* also somehow indicate that the `NoBlob` version is coming from the older - and soon-to-be phased out - design.

File: crates/bioauth-flow-rpc/src/lib.rs
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/lib.rs:1]: I think we'd want to restructure this crate as it grows; I'm thinking having a mod per rpc calls, with the impl and an `Error` located close-by.

**Review by MOZGIII — CHANGES_REQUESTED** [review]
This is getting there; please address the comments that I left, and let's have a pair coding session to finalize and polish this up further

File: crates/bioauth-flow-rpc/src/errors/robonode_request.rs
Comment (MOZGIII) [reply crates/bioauth-flow-rpc/src/errors/robonode_request.rs:1]: See other comments I left

File: crates/bioauth-flow-rpc/src/errors/authenticate.rs (hunk 7 above)
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/errors/authenticate.rs]: Improved

File: crates/bioauth-flow-rpc/src/errors/mod.rs (hunk 8 above)
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/errors/mod.rs]: Improved

File: crates/bioauth-flow-rpc/src/errors/robonode_request.rs (hunk 9 above)
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/errors/robonode_request.rs]: Improved

File: crates/bioauth-flow-rpc/src/errors/robonode_request.rs (hunk 10 above)
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/errors/robonode_request.rs]: Improved

File: crates/bioauth-flow-rpc/src/errors/enroll_v2.rs (hunk 11 above)
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/errors/enroll_v2.rs]: Done

File: crates/bioauth-flow-rpc/src/lib.rs
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/lib.rs:1]: I think we can do it during a pair coding session ?

File: crates/robonode-client/src/authenticate.rs (hunk 12 above)
Comment (sesh92) [reply crates/robonode-client/src/authenticate.rs:59]: I do, but PR is still in draft and so far to be completed, FYI I'm identifying the "kind" of error by message and also checking error data.

Comment (MOZGIII) [reply crates/robonode-client/src/authenticate.rs:59]: This was already pre-discussed with @sesh92

File: crates/bioauth-flow-rpc/src/errors/authenticate_v2.rs [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+
+    #[test]
+    fn error_key_extraction_missing_validator_key() {
+        let error: jsonrpsee::core::Error = Error(FlowBaseError::KeyExtraction(
+            ValidatorKeyError::MissingValidatorKey,
+        ))
+        .into();
+        let error: ErrorObject = error.into();
+
+        let expected_error_message = "{\"code\":500,\"message\":\"validator key not available\",\"data\":{\"validatorKeyNotAvailable\":true}}";
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/errors/authenticate_v2.rs:81]: @sesh92 check this out - we might want to add handling for this as well

File: crates/bioauth-flow-rpc/src/errors/enroll.rs [hunk 14]
```diff
@@ -2,47 +2,44 @@
...
 #[derive(Debug)]
-pub enum Error {
-    /// An error that can occur during validator key extraction.
-    KeyExtraction(ValidatorKeyError),
-    /// An error that can occur during doing a call into robonode.
-    Robonode(robonode_client::Error<robonode_client::EnrollError>),
-    /// An error that can occur during signing process.
-    Sign(SignError),
-}
+pub struct Error(pub FlowBaseError<robonode_client::EnrollError>);
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/errors/enroll.rs:10]:
```suggestion
pub struct Error(pub shared::FlowBaseError<robonode_client::EnrollError>);
```

File: crates/bioauth-flow-rpc/src/errors/enroll.rs [hunk 15]
```diff
@@ -2,47 +2,44 @@
 
 use rpc_validator_key_logic::Error as ValidatorKeyError;
 
-use super::{api_error_code, sign::Error as SignError};
+use super::{api_error_code, shared::FlowBaseError};
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/errors/enroll.rs:5]:
```suggestion
use super::{api_error_code, shared};
```

File: crates/bioauth-flow-rpc/src/lib.rs [hunk 16]
```diff
@@ -309,49 +406,34 @@ where
...
+        self.deny_unsafe.check_if_safe()?;
+
+        let EnrollResponse { scan_result_blob } = self
+            .do_enroll(liveness_data)
             .await
-            .map_err(EnrollError::Robonode)?;
+            .map_err(EnrollV2Error)?;
 
-        info!("Bioauth flow - enrolling complete");
+        let scan_result_blob = scan_result_blob.expect("expect scan result blob in response");
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/lib.rs:422]: Avoid panics or error after the action has been successfully conducted

File: crates/bioauth-flow-rpc/src/lib.rs [hunk 17]
```diff
@@ -448,4 +530,23 @@ where
...
+
+        let AuthenticateResponse {
+            auth_ticket,
+            auth_ticket_signature,
+            scan_result_blob,
+        } = self.do_authenticate(liveness_data).await.map_err(AuthenticateV2Error)?;
+
+        info!(message = "We've obtained an auth ticket", auth_ticket = ?auth_ticket);
+
+        let scan_result_blob = scan_result_blob.expect("expect scan result blob in response");
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/lib.rs:548]: See https://github.com/humanode-network/humanode/pull/1026/files#r1736850323

File: crates/bioauth-flow-rpc/src/errors/enroll.rs (hunk 15 above)
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/errors/enroll.rs]: https://github.com/humanode-network/humanode/pull/1026/commits/a761e751b8f090de9bdf7d22ae47537d7aa01749

File: crates/bioauth-flow-rpc/src/errors/enroll.rs (hunk 14 above)
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/errors/enroll.rs]: https://github.com/humanode-network/humanode/pull/1026/commits/a761e751b8f090de9bdf7d22ae47537d7aa01749

File: crates/bioauth-flow-rpc/src/errors/authenticate.rs [hunk 18]
```diff
@@ -4,18 +4,14 @@ use rpc_validator_key_logic::Error as ValidatorKeyError;
...
-pub enum Error<TxPoolError: sc_transaction_pool_api::error::IntoPoolError> {
-    /// An error that can occur during validator key extraction.
-    KeyExtraction(ValidatorKeyError),
-    /// An error that can occur during signing process.
-    Sign(SignError),
-    /// An error that can occur during doing a call into robonode.
-    Robonode(robonode_client::Error<robonode_client::AuthenticateError>),
+pub enum Error<TxPoolError> {
+    /// An error that can occur during doing a request to robonode.
+    RobonodeRequest(FlowBaseError<robonode_client::AuthenticateError>),
```
[the remark this answers is not in the document]
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/errors/authenticate.rs]: https://github.com/humanode-network/humanode/pull/1026/commits/a761e751b8f090de9bdf7d22ae47537d7aa01749

File: crates/bioauth-flow-rpc/src/lib.rs (hunk 17 above)
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/lib.rs]: https://github.com/humanode-network/humanode/pull/1026/commits/9738337651606636a4761d190a4bc93aee37967d

File: crates/bioauth-flow-rpc/src/lib.rs (hunk 16 above)
Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/lib.rs]: https://github.com/humanode-network/humanode/pull/1026/commits/9738337651606636a4761d190a4bc93aee37967d

**Review by MOZGIII — APPROVED** [review]
Looks good so far; we'd want to reconcile this once again with @sesh92 to finalize and double-check the interface correctness and that the sides match; there were some outstanding issues on the frontend side that may or may not be fixed by now on the topic of the error handling - but the happy paths should be fine now I guess

Let's hold on with the merge for now though; we'll do the refactor later once this PR merges - as to reduce the churn and weight on this works already slow progress

**Comment by sesh92** [comment]
> Looks good so far; we'd want to reconcile this once again with @sesh92 to finalize and double-check the interface correctness and that the sides match; there were some outstanding issues on the frontend side that may or may not be fixed by now on the topic of the error handling - but the happy paths should be fine now I guess
> 
> Let's hold on with the merge for now though; we'll do the refactor later once this PR merges - as to reduce the churn and weight on this works already slow progress

Can we update the [list](https://www.notion.so/humanode-network/Error-Handling-6ff497cbec3e4bb3b7ec22aac09e51cc?pvs=4) and complete it for each case?

File: crates/bioauth-flow-rpc/src/errors/authenticate_v2.rs [hunk 19]
```diff
@@ -0,0 +1,163 @@
...
+            shared::FlowBaseError::RobonodeClient(
+                err @ robonode_client::Error::Call(
+                    robonode_client::AuthenticateError::FaceScanRejectedNoBlob,
+                ),
+            ) => rpc_error_response::data(
+                api_error_code::ROBONODE,
+                err.to_string(),
+                error_data::ShouldRetry,
+            ),
+            shared::FlowBaseError::RobonodeClient(err) => {
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/errors/authenticate_v2.rs:52]: I think we'd want to unpack the robonode errors explicitly here, instead of going for a wildcard match; I'm talking about enumeration all possible `err @ robonode_client::Error` variants...

Comment (dmitrylavrenov) [reply crates/bioauth-flow-rpc/src/errors/authenticate_v2.rs]: https://github.com/humanode-network/humanode/pull/1026/commits/4bc57985fe04d89aae419caf9187889d1b0d30ac

File: crates/bioauth-flow-rpc/src/errors/enroll.rs [hunk 20]
```diff
@@ -2,47 +2,44 @@
...
 #[derive(Debug)]
-pub enum Error {
-    /// An error that can occur during validator key extraction.
-    KeyExtraction(ValidatorKeyError),
-    /// An error that can occur during doing a call into robonode.
-    Robonode(robonode_client::Error<robonode_client::EnrollError>),
-    /// An error that can occur during signing process.
-    Sign(SignError),
-}
+pub struct Error(pub shared::FlowBaseError<robonode_client::EnrollError>);
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/errors/enroll.rs:10]:
```suggestion
pub type Error = shared::FlowBaseError<robonode_client::EnrollError>;
```

File: crates/bioauth-flow-rpc/src/errors/authenticate_v2.rs [hunk 21]
```diff
@@ -0,0 +1,140 @@
+//! The `authenticate_v2` method error.
+
+use super::shared;
+use crate::error_data::{self};
```
Comment (MOZGIII) [inline crates/bioauth-flow-rpc/src/errors/authenticate_v2.rs:4]:
```suggestion
use crate::error_data;
```

**Review by MOZGIII — APPROVED** [review]

**Review by MOZGIII — CHANGES_REQUESTED** [review]
Looks like the new peer fails to work with the old robonode; more specifically, I saw enroll call failed in tests with a `reqwest` error that tries to parse empty body as JSON

File: crates/robonode-client/src/enroll.rs [hunk 22]
```diff
@@ -118,11 +157,17 @@ mod tests {
             public_key: b"123",
             liveness_data_signature: b"signature",
         };
+        let sample_response = serde_json::json!({
+            "scanResultBlob": "scanResultBlob"
+        });
+
+        let expected_response: EnrollResponse =
```
Comment (MOZGIII) [inline crates/robonode-client/src/enroll.rs:164]: This change in the tests actually masks that the new code fails to work with the old robonode; instead, add a new test

Comment (dmitrylavrenov) [reply crates/robonode-client/src/enroll.rs]: Fixed - https://github.com/humanode-network/humanode/pull/1026/commits/77f2c1057bef108e7113b68d244e68307528ba9c

File: crates/robonode-client/src/enroll.rs [hunk 23]
```diff
@@ -1,17 +1,28 @@
...
-            StatusCode::CREATED => Ok(()),
+            StatusCode::CREATED => {
+                if let Some(0) = res.content_length() {
+                    Ok(EnrollResponse {
+                        scan_result_blob: None,
+                    })
+                } else {
+                    Ok(res.json().await?)
+                }
+            }
```
Comment (MOZGIII) [inline crates/robonode-client/src/enroll.rs:25]:
```suggestion
            StatusCode::CREATED if let Some(0) = res.content_length() => Ok(EnrollResponse {
                    scan_result_blob: None,
            })
            StatusCode::CREATED => Ok(res.json().await?),
```

A slightly more compact way to represent the same logic

File: crates/robonode-client/src/enroll.rs
[the remark this answers is not in the document]
Comment (dmitrylavrenov) [reply crates/robonode-client/src/enroll.rs:1]: https://github.com/humanode-network/humanode/pull/1026/commits/a240fb477601465dd36d3823ae7afac37d301616

File: crates/robonode-client/src/enroll.rs (hunk 22 above)
Comment (dmitrylavrenov) [reply crates/robonode-client/src/enroll.rs]: https://github.com/humanode-network/humanode/pull/1026/commits/a240fb477601465dd36d3823ae7afac37d301616

File: crates/robonode-client/src/enroll.rs [hunk 24]
```diff
@@ -280,7 +323,7 @@ mod tests {
             };
 
             let response =
-                ResponseTemplate::new(case.0).set_body_json(mkerr_containing_blob(case.1, "blob"));
+                ResponseTemplate::new(case.0).set_body_json(mkerr(case.1, "scan result blob"));
```
Comment (MOZGIII) [inline crates/robonode-client/src/enroll.rs:326]: It is possible that the current protocol has replies that don't have a blob, right?
In this regard, we should specify whether there is a blob or not in the response on a case-by-case basis.

To elaborate: *this* test fn should cover cases for both `FaceScanRejected` and `InvalidPublicKey`, but only `FaceScanRejected` has a blob in the response, while `InvalidPublicKey` has no blob.

Comment (MOZGIII) [reply crates/robonode-client/src/enroll.rs:326]: On the implementation note for this: we should add a new enum `ResponseIncludesBlob { Yes, No }` and add a corresponding value of it to each case's tuple. I.e. a `bool`.

**Review by MOZGIII — APPROVED** [review]

**Comment by MOZGIII** [comment]
Great, we can now proceed with merging and shipping this as usual - to internal testing first, and then on.

**Review by dmitrylavrenov — COMMENTED** [review]
LGTM!
